### PR TITLE
Provide parsed metadata from first pass as context to second pass

### PR DESCRIPTION
**Pull request by stuarteberg**
**Note:** The "Files changed" tab here on github makes this PR seem bigger than it really is.  If you review these commits individually, you'll see that each one is small.

---

conda-build defines a number of [environment variables][1] that can be used within `build.sh`.  Using jinja templates, you can also use *most* these variables within `meta.yaml`.  I say *most* because some of those variables come from `meta.yaml` itself, such as `PKG_BUILDNUM`.

For instance, this recipe attempts to include the build number as part of the build string, but it won't work:

```
build:
  number: 99
  string: special_build_n{{PKG_BUILDNUM}}
```

We can fix this.  Since conda-build parses `meta.yaml` in *two passes*, there's no reason we can't make those variables available in the second pass.  That way, there is no longer a discrepancy between the set of variables available in `meta.yaml` and those available in `build.sh`.

It's a pretty easy fix: Just provide the `MetaData` contents (after the first pass) to the jinja context of the second pass.

[1]: http://conda.pydata.org/docs/building/environment-vars.html

--
**Edit:**

I can think of another potential use for this feature.  In `meta.yaml` the [`build: script_env:` section][2] allows the user to provide a list of "inherited" environment variables, which can be used in `build.sh`.  With this PR, those same variables can also be used within `meta.yaml` itself!  For example:

[2]: http://conda.pydata.org/docs/building/environment-vars.html#inherited-environment-variables

```
build:
  script_env:
   - USER
  string: {{USER}}_{{PKG_BUILDNUM}}
```

...ok, I'm not sure how useful that is; it just seemed worth mentioning.  The main thing I'm interested in is still the `PKG_BUILDNUM` syntax shown above.

### CODE REVIEW

**Comment by jakirkham** [comment]
Cool, thanks for drawing my attention to this.

**Comment by stuarteberg** [comment]
So... not trying to be annoying, I'm just playing the role of "squeaky wheel" for my final jinja-related PR here.

Since you guys (@msarahan, @groutr) have been merging jinja-related PRs lately and you have this stuff fresh in your mind, can you take a look at this last one?  I added a test!

File: tests/test-recipes/metadata/jinja_vars/build.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# CONDA_TEST_VAR was inherited via build/script_env
+[ "${CONDA_TEST_VAR}" != "" ]
```
Comment (pelson) [inline tests/test-recipes/metadata/jinja_vars/build.sh]: Maybe it would be healthy to be more assertive about this. (i.e. ```== "conda_test"```).

File: conda_build/metadata.py [hunk 2]
```diff
@@ -605,6 +543,72 @@ def binary_has_prefix_files(self):
...
+        # search relative to current conda environment directory
+        conda_env_path = os.environ.get('CONDA_DEFAULT_ENV')  # path to current conda environment
+        if conda_env_path and os.path.isdir(conda_env_path):
+            conda_env_path = os.path.abspath(conda_env_path)
+            conda_env_path = conda_env_path.replace('\\', '/') # need unix-style path
+            env_loader = jinja2.FileSystemLoader(conda_env_path)
+            loaders.append(jinja2.PrefixLoader({'$CONDA_DEFAULT_ENV': env_loader}))
+    
+        undefined_type = jinja2.StrictUndefined
+        if permit_undefined_jinja:
```
Comment (pelson) [inline conda_build/metadata.py]: I'd love to have a ton of small unit tests for this. Is seems like it is a powerful part of the implementation, and I'd be more comfortable if I had a test which actually exercised the important parts (for instance, what would happen if I removed the ``__int__`` strangeness...)

Comment (pelson) [inline conda_build/metadata.py]: (I see you didn't actually write this bit, so sorry for dumping that suggestion on you!)

File: conda_build/environ.py [hunk 3]
```diff
@@ -144,6 +144,7 @@ def get_dict(m=None, prefix=None):
         d['PKG_NAME'] = m.name()
         d['PKG_VERSION'] = m.version()
         d['PKG_BUILDNUM'] = str(m.build_number())
+        d['PKG_BUILD_STRING'] = str(m.build_id())
```
Comment (pelson) [inline conda_build/environ.py]: Maybe we could just pass the meta itself?

File: conda_build/metadata.py (hunk 2 above)
Comment (stuarteberg) [inline conda_build/metadata.py]: >I see you didn't actually write this bit

Right, this isn't part of the current PR -- it was just relocated from elsewhere in the file.  I did not write the lines above (related to jinja loaders).  But as luck would have it, I did write the `UndefinedNeverFail` class below (in an earlier PR).

>for instance, what would happen if I removed the `__int__` strangeness...

Without the `__int__ = lambda _: 0` line, templates like `{{ int(PKG_VERSION) }}`, would raise an error during the first pass (since `PKG_VERSION` isn't defined until the second pass).  (Side note: I see `__int__` is accidentally set twice in this code... no harm there, but I'll fix it now.)

File: tests/test-recipes/metadata/jinja_vars/build.sh (hunk 1 above)
Comment (stuarteberg) [inline tests/test-recipes/metadata/jinja_vars/build.sh]: Good idea.  Done.

File: conda_build/environ.py (hunk 3 above)
Comment (stuarteberg) [inline conda_build/environ.py]: >Maybe we could just pass the meta itself?

I'm not sure I understand the question.  You mean something like this?

```
d['PKG_BUILD_STRING'] = m.get_value('build/string', '')
```

I think `build_id()` is preferable because it provides a default value in some cases.

Comment (pelson) [inline conda_build/environ.py]: No, I meant:

```
d['recipe'] = m
```

Within Jinja, we should be able to do things like:

```
{{ recipe.name() }} 
```

right?

Comment (stuarteberg) [inline conda_build/environ.py]: Oh, that's an interesting idea.  It turns out this isn't the right function to do it (the output of `environ.get_dict()` must be strings only), but we could add these two lines to `jinja_context.context_processor()`:

```python
    if initial_metadata:
        ctx['metadata'] = initial_metadata
```

...which does the trick.  Well, we also need one more minor change needed in our special `UndefinedNeverFail` class:

```python
class UndefinedNeverFail(jinja2.Undefined):

    # ...
    
    def __getattr__(self, k):
        try:
            return object.__getattr__(k)
        except AttributeError:
            return UndefinedNeverFail(name=self._undefined_name + '.' + k)
```

...and then your idea works perfectly.  For example, this recipe does what you expect:

```yaml
package:
  name: conda-build-test-jinja-vars
  version: 1.0

build:
  number: 123
  string: {{CONDA_TEST_VAR}}_{{PKG_BUILDNUM}}

  script_env:
    - CONDA_TEST_VAR
  
about:
  summary: {{ metadata.name() }} is just a test package.
```

I'm intrigued by this idea, but I'm also wary of it.  This change would encourage recipe authors to depend explicitly on imlementation details within `conda-build`. 

At the moment, recipes are limited using to a predefined, documented set of variables (e.g. `PKG_BUILDNUM`).  No matter what changes we make to `conda-build` in the future, it will be easy to avoid breaking backward compatibility with old recipes.

But if we expose the `MetaData` class directly to the recipe, then simple changes to `conda-build` (say, renaming `build_id()` to `build_string()`) have the potential to break recipes in the wild, if those recipes use `{{ metadata.build_id() }}`.

In conclusion: I tried your idea on my machine, and it works.  But after some thought, I don't want to add it to this PR.

Comment (msarahan) [inline conda_build/environ.py]: +1 for maintainability at the cost of some usability.  This would be a nice thing, but the potential for breaking backwards compatibility is huge.

**Comment by stuarteberg** [comment]
So.... any verdict here?

**Comment by msarahan** [comment]
Thanks for your patience, and for you hard work on this PR.
